Fix default encryption type

bndw opened the pull request:
Related to https://github.com/bndw/pick/issues/186

bndw committed:
Fix default encryption type

Related to https://github.com/bndw/pick/issues/186

leonklingele reviewed (approved):
👍 Some (basic) tests would be nice as well so we don't have to deal with this again in the future.

bndw commented:
Do you have ideas for a test case? We're currently relying on the [default behavior](https://github.com/bndw/pick/blob/master/crypto/client.go#L19-L23) of the `crypto.New` method to dictate encryption type, when it's not specified.

leonklingele commented:
Off the top of my head (untested):

```go
func TestNew(t *testing.T) {
	fixtures := []struct{
		t string
		fn func(Client) bool
	}{
		{
			t: "",
			fn := func(c Client) bool {
				_, ok := c.(ChaCha20Poly1305Client)
				return ok
			},
		},
		{
			t: "blubb",
			fn := func(c Client) bool {
				_, ok := c.(ChaCha20Poly1305Client)
				return ok
			},
		},
		{
			t: ConfigTypeChaChaPoly,
			fn := func(c Client) bool {
				_, ok := c.(ChaCha20Poly1305Client)
				return ok
			},
		},
		{
			t: ConfigTypeOpenPGP,
			fn := func(c Client) bool {
				_, ok := c.(OpenPGPClient)
				return ok
			},
		},
		{
			t: ConfigTypeAESGCM,
			fn := func(c Client) bool {
				_, ok := c.(AESGCMClient)
				return ok
			},
		},
	}

	for _, f := range fixtures {
		t.Run(f.t, func(t *testing.T) {
			conf := NewDefaultConfig()
			conf.Type = f.t
			if c, err := New(conf); err != nil {
				t.Fatal(err)
			} else {
				if !f.fn(c) {
					t.Fatalf("failed to type assert")
				}
			}
		})
	}
}
```

bndw commented:
I was curious if there was a way to test the `fallthrough` behavior we're depending on, but I'm not sure that's possible/worth the effort.

I'll add some simple blackbox tests like you've referenced above, thanks.

bndw committed:
Add tests

bndw committed:
Remove testing with < Gov1.9

Our tests depend on atleast v1.7 and 2 versions prior (1.11 current) seems sufficient

bndw committed:
TravisCI needs versions wrapped in quotes

Related to https://github.com/travis-ci/travis-ci/issues/9247

bndw commented:
`go tool vet` on tip is having issues, causing this build to fail. Failures on tip should probably be warnings, rather than blockers, since it's unstable.

https://github.com/golang/go/issues/28869

bndw committed:
Allow golang tip failures

bndw commented:
I've updated the TravisCI config to get these tests passing. I'd have preferred to do that in a different PR, but the failed checks here got the best of me.

- Tests Go v1.9.x to the latest release, including tip
- Allows failures of Go tip

leonklingele commented on crypto/client_test.go:
```diff
@@ -0,0 +1,63 @@
+			},
+		},
+		{
+			t: "chachapoly",
```
Why not use `ConfigTypeChaChaPoly` (applies to all `t`s below as well)?

bndw replied:
Yeah, using the consts would make a lot of sense 😆

leonklingele commented:
I feel like we should also add a test to ensure that a serialized safe is stored using `"type":"chachapoly"` and not something else.

bndw committed:
Use const values for encryption type

leonklingele commented on crypto/client_test.go:
```diff
@@ -0,0 +1,63 @@
+	}{
+		{
+			t: "",
+			fn: func(c interface{}) bool {
```
Hmm, why did you decide to accept an empty interface here instead of a `Client` type? Applies to lines below too.

bndw replied:
No [good] reason. Updated, thanks!

bndw committed:
Use Client interface in tests, instead of the generic interface{}

bndw committed:
Test that the default safe encryption is chachapoly in serialization

bndw commented:
@leonklingele I added a quick and dirty test for asserting that chachapoly is used for default encryption during serialization. Most of that code could be refactored with `./safe/utils_test.go` at some later point.

leonklingele commented on safe/save_test.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	// Assert the default safe encryption is ChaChaPoly
+	if dto.Config.ChaCha20Poly1305Settings == nil {
```
Is `dto.Config.Type` not populated when loading a safe? If it is, it should be checked too just to be sure.

bndw replied:
It's not; that was my first test and it's an empty string. We may consider populating that in the `crypto.New` implementation, but I felt that was out of scope for this PR.

leonklingele reviewed (approved):
👍

leonklingele commented:
This PR makes it even easier to switch to a different default encryption type: https://github.com/bndw/pick/pull/188
